### DISCUSSION TITLE
Proposal: Multiple possibilities return value

### DISCUSSION THREAD

**Discussion by hez2010** [dotnet/csharplang#1891] [Language Ideas]
If I want a function to return value of different types, I have to write like this: 

```c#
public void m()  // can constraint the type of returned value but have to write a lot of duplicated code
{
     if (....) 
     {
         int x;
         x = a();
     }
     else 
     {
         string x;
         x = a();
     }
}
public int a()
{
      return 5;
}
public string a()
{
      return "t";
}
```

or 

```c#
public object a() // cannot constraint the type of returned value
{
      if (...) return 5;
      else return "t";
}
```

I think the code below will be useful and can both constraint the type of returned value and reduce duplicated code:

```c#
public string|int a()
{
      if (...) return 5;
      else return "t";
}
```

**Comment by HaloFour** [comment]
So, if you have `var value = a()`, what is the type of `value`?  How can you use it?

**Comment by DavidArno** [comment]
What you are proposing is often implemented as Discriminated Unions, which is championed in #113.

**Comment by jnm2** [comment]
Sounds like https://github.com/dotnet/csharplang/issues/399.

I have a `OneOf<T1, T2>` type I use for ad-hoc discriminated unions (the way ValueTuple is an ad-hoc product type).

**Comment by hez2010** [comment]
@HaloFour 
The type of value is object when the variable is implicitly declared and there's no difference between object and T1|T2|....

**Comment by Happypig375** [comment]
Syntax 1:
It has a lot of problems. If the type when `var` is used is `object`, will the overload returning an `int` or the one returning an `string` be called?

Syntax 2: It is already valid today. What are you proposing?

Syntax 3: Duplicate of #399.

**Comment by theunrepentantgeek** [comment]
This proposes a TypeScript idiom that makes a lot of sense in the dynamically typed world of JavaScript; it doesn't make any sense to me in C#.

Shapes and Extensions (#164) would provide all of the benefits of this proposal, and a lot more, but in a form that's type safe and rigorous, as suits the nature of C#.

**Comment by Austin-bryan** [comment]
@hez2010 From what I understand, you can't overload on return types for very good reasons. It had to do with the compiler never being able to know which when they meant. All of these examples use your first syntax example. Examples:

```csharp
a();    // does string or int get called?

...

public void Foo(int ...) { ... }
public void Foo(string s) { ... }
...
Foo(a())    // which overload is called and which a is called?

...

object mysteryType = a();    // Which a gets called? 
int x = mysteryType;    // ...and will this work? It depends, is mysteryType an int or string? I don't know and neither does the compiler
```

etc., I'm probably missing a lot here. 

In the meantime, just use `dynamic`, it's basically the same thing except that dynamic can also be `Foo`, but it'll still get your job done.

**Comment by alexgman** [comment]
> @hez2010 From what I understand, you can't overload on return types for very good reasons. It had to do with the compiler never being able to know which when they meant. All of these examples use your first syntax example. Examples:
> 
> ```cs
> a();    // does string or int get called?
> 
> ...
> 
> public void Foo(int ...) { ... }
> public void Foo(string s) { ... }
> ...
> Foo(a())    // which overload is called and which a is called?
> 
> ...
> 
> object mysteryType = a();    // Which a gets called? 
> int x = mysteryType;    // ...and will this work? It depends, is mysteryType an int or string? I don't know and neither does the compiler
> ```
> 
> etc., I'm probably missing a lot here.
> 
> In the meantime, just use `dynamic`, it's basically the same thing except that dynamic can also be `Foo`, but it'll still get your job done.

can you please give an example of how to use `dynamic` as a union type solution? for example, if i want to have a discriminated union of "mystring1" and "mystring2" how do i do that with `dynamic`?

**Comment by DarthVella** [comment]
The ```dynamic``` type/keyword is used to force type and member checking of objects to be deferred until runtime and can be used to implement crude, unchecked type unions (indiscriminate unions?) between unrelated objects with similarly named properties and methods. It cannot be used to create an enforced union of string values as can be done in Typescript. If that's what you're after, I'd suggest creating an enum and calling the ToString method on it.
